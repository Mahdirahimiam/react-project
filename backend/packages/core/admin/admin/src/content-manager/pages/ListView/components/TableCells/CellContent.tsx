import { Tooltip, Typography } from '@strapi/design-system';
import isEmpty from 'lodash/isEmpty';
import styled from 'styled-components';

import { isFieldTypeNumber } from '../../../../utils/fields';

import { CellValue } from './CellValue';
import { SingleComponent, RepeatableComponent } from './Components';
import { MediaSingle, MediaMultiple } from './Media';
import { RelationSingle, RelationMultiple } from './Relations';

import type { FormattedContentTypeLayout } from '../../../../utils/layouts';
import type { TableHeader } from '../../ListViewPage';
import type { Attribute, Entity } from '@strapi/types';

interface CellContentProps extends Omit<TableHeader, 'key'> {
  content: Attribute.GetValue<Attribute.Any>;
  rowId: Entity.ID;
  contentType: FormattedContentTypeLayout;
}

const CellContent = ({
  content,
  fieldSchema,
  metadatas,
  name,
  rowId,
  contentType,
}: CellContentProps) => {
  if (!hasContent(content, metadatas, fieldSchema)) {
    return <Typography textColor="neutral800">-</Typography>;
  }

  switch (fieldSchema.type) {
    case 'media':
      if (!fieldSchema.multiple) {
        return <MediaSingle {...content} />;
      }

      return <MediaMultiple content={content} />;

    case 'relation': {
      if (isSingleRelation(fieldSchema.relation)) {
        return <RelationSingle metadatas={metadatas} content={content} />;
      }

      return (
        <RelationMultiple
          metadatas={metadatas}
          content={content}
          name={name}
          entityId={rowId}
          uid={contentType.uid}
        />
      );
    }

    case 'component':
      if (fieldSchema.repeatable === true) {
        return <RepeatableComponent content={content} metadatas={metadatas} />;
      }

      return <SingleComponent content={content} metadatas={metadatas} />;

    case 'string':
      return (
        <Tooltip description={content}>
          <TypographyMaxWidth ellipsis textColor="neutral800">
            <CellValue type={fieldSchema.type} value={content} />
          </TypographyMaxWidth>
        </Tooltip>
      );

    default:
      return (
        <TypographyMaxWidth ellipsis textColor="neutral800">
          <CellValue type={fieldSchema.type} value={content} />
        </TypographyMaxWidth>
      );
  }
};

const TypographyMaxWidth = styled(Typography)`
  max-width: 300px;
`;

const hasContent = (
  content: CellContentProps['content'],
  metadatas: CellContentProps['metadatas'],
  fieldSchema: CellContentProps['fieldSchema']
) => {
  if (fieldSchema.type === 'component') {
    const { mainField } = metadatas;

    // Repeatable fields show the ID as fallback, in case the mainField
    // doesn't have any content
    if (fieldSchema?.repeatable || !mainField) {
      return content && Array.isArray(content) && content.length > 0;
    }

    const value = content?.[mainField.name];

    // relations, media ... show the id as fallback
    if (mainField.name === 'id' && ![undefined, null].includes(value)) {
      return true;
    }

    /* The ID field reports itself as type `integer`, which makes it
       impossible to distinguish it from other number fields.

       Biginteger fields need to be treated as strings, as `isNumber`
       doesn't deal with them.
    */
    if (
      isFieldTypeNumber(mainField.type) &&
      mainField.type !== 'biginteger' &&
      mainField.name !== 'id'
    ) {
      return typeof value === 'number';
    }

    return !isEmpty(value);
  }

  if (fieldSchema.type === 'relation') {
    if (isSingleRelation(fieldSchema.relation)) {
      return !isEmpty(content);
    }

    return content?.count > 0;
  }

  /*
      Biginteger fields need to be treated as strings, as `isNumber`
      doesn't deal with them.
  */
  if (isFieldTypeNumber(fieldSchema.type) && fieldSchema.type !== 'biginteger') {
    return typeof content === 'number';
  }

  if (fieldSchema.type === 'boolean') {
    return content !== null;
  }

  return !isEmpty(content);
};

const isSingleRelation = (
  type: Extract<CellContentProps['fieldSchema'], { type: 'relation' }>['relation']
) => ['oneToOne', 'manyToOne', 'oneToOneMorph'].includes(type);

export { CellContent };
export type { CellContentProps };
