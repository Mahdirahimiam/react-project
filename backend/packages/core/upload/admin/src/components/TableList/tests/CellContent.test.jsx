import React from 'react';

import { lightTheme, ThemeProvider } from '@strapi/design-system';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import { CellContent } from '../CellContent';

const PROPS_FIXTURE = {
  cellType: 'image',
  contentType: 'asset',
  content: {
    alternativeText: 'alternative alt',
    ext: 'jpeg',
    formats: {
      thumbnail: {
        url: 'michka-picture-url-thumbnail.jpeg',
      },
    },
    mime: 'image/jpeg',
    url: 'michka-picture-url-default.jpeg',
  },
  name: 'preview',
};

const ComponentFixture = (props) => {
  const customProps = {
    ...PROPS_FIXTURE,
    ...props,
  };

  return (
    <IntlProvider locale="en" messages={{}}>
      <ThemeProvider theme={lightTheme}>
        <CellContent {...PROPS_FIXTURE} {...customProps} />
      </ThemeProvider>
    </IntlProvider>
  );
};

const setup = (props) => render(<ComponentFixture {...props} />);

describe('TableList | CellContent', () => {
  it('should render image cell type when element type is asset and mime includes image', () => {
    const { container, getByRole } = setup();

    expect(getByRole('img', { name: 'alternative alt' })).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render image cell type when element type is asset and mime does not include image', () => {
    const { container, getByText } = setup({
      content: { ...PROPS_FIXTURE.element, mime: 'application/pdf', ext: 'pdf' },
    });

    expect(getByText('pdf')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render image cell type when element type is folder', () => {
    const { container } = setup({ contentType: 'folder' });

    expect(container.querySelector('path')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render text cell type', () => {
    const { container, getByText } = setup({
      cellType: 'text',
      content: { ...PROPS_FIXTURE.content, name: 'some text' },
      name: 'name',
    });

    expect(getByText('some text')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render extension cell type when element type is asset', () => {
    const { container, getByText } = setup({
      cellType: 'ext',
      content: { ...PROPS_FIXTURE.content, ext: '.pdf' },
      name: 'ext',
    });

    expect(getByText('PDF')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render extension cell type with "-" when element type is folder', () => {
    const { container, getByText } = setup({ cellType: 'ext', contentType: 'folder' });

    expect(getByText('-')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render size cell type when element type is asset', () => {
    const { container, getByText } = setup({
      cellType: 'size',
      content: { ...PROPS_FIXTURE.content, size: '20.5435' },
      name: 'size',
    });

    expect(getByText('21KB')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render size cell type with "-" when element type is folder', () => {
    const { container, getByText } = setup({ cellType: 'size', contentType: 'folder' });

    expect(getByText('-')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render date cell type', () => {
    const { container, getByText } = setup({
      cellType: 'date',
      content: { ...PROPS_FIXTURE.content, updatedAt: '2022-11-18T12:08:02.202Z' },
      name: 'updatedAt',
    });

    expect(getByText('Friday, November 18, 2022')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render "-" by default when no recognized cell type is passed', () => {
    const { container, getByText } = setup({ cellType: 'not recognized type' });

    expect(getByText('-')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
