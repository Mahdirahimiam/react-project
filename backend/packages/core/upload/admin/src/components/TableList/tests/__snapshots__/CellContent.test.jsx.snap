// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableList | CellContent should render "-" by default when no recognized cell type is passed 1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
}

<div>
  <span
    aria-label="This field is empty"
    class="c0"
  >
    -
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render date cell type 1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
}

<div>
  <span
    class="c0"
  >
    Friday, November 18, 2022
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render extension cell type when element type is asset 1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
}

<div>
  <span
    class="c0"
  >
    PDF
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render extension cell type with "-" when element type is folder 1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
}

<div>
  <span
    aria-label="This field is empty"
    class="c0"
  >
    -
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render image cell type when element type is asset and mime does not include image 1`] = `
.c4 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c3 {
  font-size: 0.875rem;
  line-height: 1.43;
  font-weight: 600;
  color: #0c75af;
  text-transform: uppercase;
}

.c0 {
  background: #eaf5ff;
  border-radius: 50%;
  width: 32px;
  height: 32px;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 span {
  line-height: 0;
}

<div>
  <div
    class="c0 c1 c2"
  >
    <span
      class="c3"
    >
      pdf
    </span>
  </div>
  <div
    class="c4"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render image cell type when element type is asset and mime includes image 1`] = `
.c2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  position: relative;
  width: 32px;
  height: 32px;
}

.c1 {
  border-radius: 50%;
  object-fit: cover;
  display: block;
  position: relative;
}

<div>
  <span>
    <div
      class="c0"
    >
      <img
        alt="alternative alt"
        class="c1"
        height="32px"
        src="michka-picture-url-thumbnail.jpeg"
        width="32px"
      />
    </div>
  </span>
  <div
    class="c2"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render image cell type when element type is folder 1`] = `
.c6 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c3 {
  font-size: 0.875rem;
  line-height: 1.43;
  font-weight: 600;
  color: #0c75af;
  text-transform: uppercase;
}

.c0 {
  background: #eaf5ff;
  border-radius: 50%;
  width: 32px;
  height: 32px;
}

.c4 {
  color: #66b7f1;
  width: 1rem;
  height: 1rem;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5 path {
  fill: #66b7f1;
}

.c2 span {
  line-height: 0;
}

<div>
  <div
    class="c0 c1 c2"
  >
    <span
      class="c3"
    >
      <svg
        class="c4 c5"
        fill="none"
        height="1rem"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.414 5H21a1 1 0 0 1 1 1v14a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h7.414l2 2Z"
          fill="#212134"
        />
      </svg>
    </span>
  </div>
  <div
    class="c6"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render size cell type when element type is asset 1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
}

<div>
  <span
    class="c0"
  >
    21KB
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render size cell type with "-" when element type is folder 1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
}

<div>
  <span
    aria-label="This field is empty"
    class="c0"
  >
    -
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`TableList | CellContent should render text cell type 1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
}

<div>
  <span
    class="c0"
  >
    some text
  </span>
  <div
    class="c1"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
